pkg/controller/contrailcni/daemonset.go

```diff
@@ -0,0 +1,206 @@
+package contrailcni
+
+import (
+	apps "k8s.io/api/apps/v1"
+	core "k8s.io/api/core/v1"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type CniDirs struct {
+	BinariesDirectory string
+	DeploymentType    string
+}
+
+//GetDaemonset returns DaemonSet object for vRouter CNI
+func GetDaemonset(cniDir CniDirs, requestName, instanceType string) *apps.DaemonSet {
+	var trueVal = true
+
+	var podIPEnv = core.EnvVar{
+		Name: "POD_IP",
+		ValueFrom: &core.EnvVarSource{
+			FieldRef: &core.ObjectFieldSelector{
+				FieldPath: "status.podIP",
+			},
+		},
+	}
+
+	var podInitContainers = []core.Container{
+		{
+			Name:  "vroutercni",
+			Image: "hub.juniper.net/contrail-nightly/contrail-kubernetes-cni-init:master.latest",
+			Command: []string{"sh", "-c",
+				"mkdir -p /host/etc_cni/net.d && " +
+					"mkdir -p /var/lib/contrail/ports/vm && " +
+					"cp -f /usr/bin/contrail-k8s-cni /host/opt_cni_bin && " +
+					"chmod 0755 /host/opt_cni_bin/contrail-k8s-cni && " +
+					"cp -f /etc/contrailconfigmaps/10-contrail.conf /host/etc_cni/net.d/10-contrail.conf && " +
+					"tar -C /host/opt_cni_bin -xzf /opt/cni-v0.3.0.tgz"},
+			Env: []core.EnvVar{
+				podIPEnv,
+			},
+			VolumeMounts: []core.VolumeMount{
+				core.VolumeMount{
+					Name:      "var-lib-contrail",
+					MountPath: "/var/lib/contrail",
+				},
+				core.VolumeMount{
+					Name:      "cni-config-files",
+					MountPath: "/host/etc_cni",
+				},
+				core.VolumeMount{
+					Name:      "cni-bin",
+					MountPath: "/host/opt_cni_bin",
+				},
+				core.VolumeMount{
+					Name:      "var-log-contrail-cni",
+					MountPath: "/host/log_cni",
+				},
+				core.VolumeMount{
+					Name:      "vrouter-logs",
+					MountPath: "/var/log/contrail",
+				},
+				core.VolumeMount{
+					Name:      requestName + "-" + instanceType + "-volume",
+					MountPath: "/etc/contrailconfigmaps",
+				},
+			},
+			ImagePullPolicy: "Always",
+			SecurityContext: &core.SecurityContext{
+				Privileged: &trueVal,
+			},
+		},
+	}
+
+	if cniDir.DeploymentType == "openshift" {
+		podInitContainers = append(podInitContainers, core.Container{
+			Name:  "multusconfig",
+			Image: "busybox",
+			Command: []string{
+				"sh",
+				"-c",
+				"mkdir -p /etc/kubernetes/cni/net.d && " +
+					"cp -f /etc/contrailconfigmaps/10-contrail.conf /etc/kubernetes/cni/net.d/10-contrail.conf && " +
+					"mkdir -p /var/run/multus/cni/net.d && " +
+					"cp -f /etc/contrailconfigmaps/10-contrail.conf /var/run/multus/cni/net.d/80-openshift-network.conf"},
+			VolumeMounts: []core.VolumeMount{
+				core.VolumeMount{
+					Name:      "etc-kubernetes-cni",
+					MountPath: "/etc/kubernetes/cni",
+				},
+				core.VolumeMount{
+					Name:      "multus-cni",
+					MountPath: "/var/run/multus",
+				},
+			},
+			ImagePullPolicy: "Always",
+		})
+	}
+
+	var podContainers = []core.Container{
```

Comment:
Please we shouldn't leave a dummy service, this should be a job, or something job like

Reply:
Unfortunately this has to run on each node that matches the specified nodeSelector, and currently there is no type with this behaviour in k8s. There are long discussions in kubernetes' github.
Here is a nice comment which lists the possible workarounds : https://github.com/kubernetes/kubernetes/issues/64623#issuecomment-609875003

On of the methods is the one Szymon chose which is to have a daemonset with the task executed by the initContainer and then a container which does nothing and never exits.

Reply:
I would try:
- job with anti-affinity might actually solve the problem
- create a custom controller for our own DaemonJob that uses daemon set but it deletes it after job is done